recipes/full_finetune_distributed.py

```diff
@@ -614,6 +671,60 @@ def _setup_data(
 
         return sampler, dataloader
 
+    def save_checkpoint_async(
+        self,
+        epoch: int,
+    ) -> None:
+        """
+        Checkpoint the state of the recipe. The constructed checkpoint state dict
+        contains the following information:
+        - Model weights with key training.MODEL_KEY
+        - Relevant recipe state if training is not complete
+
+        Checkpointer will save the model weights and recipe state in an asynchrnous way
+        which unblocks the training sooner to continue. To correctly resume training from
+        an intermediate checkpoint, user needs to have both resume_from_checkpoint and
+        enable_async_checkpointing flags set to True in the config.
+        """
+
+        if self._is_rank_zero:
+            log.info("Saving checkpoint asynchronously. Retrieving full state dict...")
+            cp_start = time.perf_counter()
+
+        # Create the checkpoint dict to be sent to the checkpointer and ultimately persisted to storage
+        ckpt_dict = {
+            training.SEED_KEY: self.seed,
+            training.EPOCHS_KEY: self.epochs_run,
```

Comment:
@felipemello1 Sorry. Could you please explain this comment a bit more? Are you suggesting, we should not be saving the epochs_run in the checkpoint? How would be restore the training progress in terms of the number of epochs that were run till the last checkpoint was taken?

Reply:
hey, sorry, let me clarify: do we need to pass "epoch" as input args AND also track it with self.epochs_run? Are those different? If they are, maybe we can rename to make obvious the difference between those

![image](https://github.com/user-attachments/assets/1d9251d2-3363-4c07-8a7b-8a107d39e4a3)

Reply:
We can do that. save_checkpoint already takes the `epoch` so I propagated it to the async API as well. It would require a minor change to update the self._epochs_run only after taking the checkpoint for the current epoch here: https://github.com/pytorch/torchtune/blob/main/recipes/full_finetune_distributed.py#L878

Reply:
@felipemello1 I looked into it some more and decided to revert my changes back to the prior state. 

- current_epoch refers to the epoch which just got completed and epochs_run gets incremented to the next epoch. 
- current_epoch is used in the checkpoint name however in the training progress state dict, intermediate checkpoints save the next epoch to resume from but final save checkpoints save the epoch which just got completed. 

Therefore passing the current_epoch in the checkpoint methods seems cleaner for now compared to doing +1s and -1s for different type of checkpoints. I do believe this logic should be refactored a bit and all the checkpoints should save the total epochs which have completed. During training run, we adjust the starting epoch depending on if we are resuming from a checkpoint and training from scratch.